akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -605,6 +605,14 @@ import scala.util.control.NonFatal
     // call has no effect and therefore does the right thing: nothing.
     try {
       inputs.foreach(_.onInternalError(reason))
+      // fail all sources
+      logics.foreach { logic ⇒
+        // source
+        if (logic.inCount == 0 && logic.outCount > 0) {
```

Comment:
Finish does not know about the exception, so it cannot fail the stages. If we'd separate the fail from the stop, AFAICS the `finish()` method has no way to differentiate between already stopped stages and stages that are "halfway" stopped because of the failure.

One thing I just realised though based on your last comment is if stages stop the propagation this does not work and we need to go through all stages and fail them as it is impossible to recover in this case.